app/models/vim_performance_state/purging.rb

```diff
@@ -8,9 +8,28 @@ def purge_mode_and_value
         %w(orphaned resource)
       end
 
+      # remove anything where the resource no longer exists AND
+      # remove anything older than a certain date
+      def purge_timer
+        purge_queue(:orphaned, "resource")
+        purge_queue(:date, purge_date)
+      end
+
       def purge_window_size
         ::Settings.vim_performance_states.history.purge_window_size
       end
+
+      #
+      # By Date
+      #
+
+      def purge_date
+        ::Settings.vim_performance_states.history.keep_states.to_i_with_method.seconds.ago.utc
+      end
+
+      def purge_scope(older_than)
+        where(arel_table[:timestamp].lt(older_than))
+      end
```

Comment:
Based on my comment above, I think this can just more succintly be:


```suggestion
      def purge_scope(older_than)
        where(:resource_id => nil).where(arel_table[:timestamp].lt(older_than))
      end
```

And then change the purge_mode_and_value to:

```diff
        def purge_mode_and_value
-         %w(orphaned resource)
-         [:scope, purge_date]
        end

Reply:
so after a discussion, we determined orphaned is saying that this record is owned by a record that was removed from the database.

In theory, I understand how this can happen, but it looked like all of the references were setup to destroy this record.

Reply:
Oh right - we discussed this and there are 2 types of orphans

1. resource_id is nil
2. resource_id is not nil but the corresponding record not longer exists

The main way this can happen is whether the caller uses dependent => :nullify (1) vs nothing at all (2).

`where(:resource_id => nil)` would only handle the first case, whereas orphaned scope handles the second case.  

BinaryBlobs in partucular does `where(:resource_id => nil).where(arel_table[:timestamp].lt(older_than))`, which _partially_ handles the first case by applying a timestamp check _AND_ and resource_id => nil check.  If I recall, we can create BinaryBlobs without a resource_id in some cases, particularly in provisioning, and there was a bug where that BinaryBlob was getting removed during the provisioning and breaking things, which is why that one is "special"

Reply:
tack on: "nothing at all"
there is no `dependent: ...` or the association is not even mentioned

Reply:
aah, came across this example:

```
class ExtManagementSystem
  has_many    :vim_performance_states, :as => :resource  # Destroy will be handled by purger
end
```